### ISSUE TITLE
MSFT_SPStateServiceApp: The specified database already exists.

### ISSUE THREAD

**Issue by VerdonTrigance** [dsccommunity/SharePointDsc#1216]
ISSUE TITLE:
MSFT_SPStateServiceApp: The specified database already exists.

    ISSUE DESCRIPTION (this template):
Trying to install new sp farm 2019 from scratch in new clear VM brings some errors after which on second run I am getting such error. Previous run successfully creates database for StateService, but not the services itself. 
Manual delete database and rerun doesn't help - it creates database and throw this error again.

#### Details of the scenario you tried and the problem that is occurring

Problem in existed database. We need to check database status too.

#### Verbose logs showing the problem

First run return this:

```
The PowerShell DSC resource '[SPFarm]CreateSPFarm' with SourceInfo 'C:\SharePoint.ps1::94::9::SPFarm' threw one or 
more non-terminating errors while running the Set-TargetResource functionality. These errors are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel 
for more details.
```

After which: 

```
VERBOSE: [dev02]: LCM:  [ Start  Resource ]  [[SPStateServiceApp]StateServiceApp]
VERBOSE: [dev02]: LCM:  [ Start  Test     ]  [[SPStateServiceApp]StateServiceApp]
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Testing for state service application State Service Application
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Getting state service application 'State Service Application'
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Executing as the local run as user dom\user
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Leaving BeginProcessing Method of Get-SPStateServiceApplication.
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Leaving ProcessRecord Method of Get-SPStateServiceApplication.
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Leaving EndProcessing Method of Get-SPStateServiceApplication.
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Current Values: DatabaseName=dev02_State; Ensure=Absent; Name=State Service Application
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Target Values: DatabaseName=dev02_State; Ensure=Present; Name=State Service Application; Verbose=
True
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Test-TargetResource returned False
VERBOSE: [dev02]: LCM:  [ End    Test     ]  [[SPStateServiceApp]StateServiceApp]  in 0.6400 seconds.
VERBOSE: [dev02]: LCM:  [ Start  Set      ]  [[SPStateServiceApp]StateServiceApp]
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Setting state service application 'State Service Application'
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Creating State Service Application State Service Application
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Executing as the local run as user dom\user
VERBOSE: [dev02]:                            [[SPStateServiceApp]StateServiceApp] Leaving BeginProcessing Method of New-SPStateServiceDatabase.
The specified database already exists.
    + CategoryInfo          : InvalidData: (Microsoft.Offic...eDatabaseCmdlet:) [], CimException
    + FullyQualifiedErrorId : Microsoft.Office.Server.Administration.NewStateServiceDatabaseCmdlet
    + PSComputerName        : dev02
```

#### Suggested solution to the issue

Additional parameter in Get-TargetResource "databaseExists" might be helpful. For example:

$serviceDatabase = Get-SPStateServiceDatabase -Identity $params.DatabaseName `
            -ErrorAction SilentlyContinue

        if ($null -ne $serviceDatabase)
        {
            $databaseExists = $true
        }

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

```powershell
SPStateServiceApp StateServiceApp
        {
            Name                 = "State Service Application"
            DatabaseName         = "dev02_State"
            PsDscRunAsCredential = $SPSetupAccount
            DependsOn            = "[SPFarm]CreateSPFarm"
        }
```

#### The operating system the target node is running

    OsName               : Microsoft Windows Server 2019 Standard
OsOperatingSystemSKU : StandardServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version of SharePoint that is used (e.g. SharePoint 2016)

SharePoint 2019

#### Version and build of PowerShell the target node is running
Name                           Value                                                                                                                                                       
----                           -----                                                                                                                                                       
PSVersion                      5.1.17763.1007                                                                                                                                              
PSEdition                      Desktop                                                                                                                                                     
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                     
BuildVersion                   10.0.17763.1007                                                                                                                                             
CLRVersion                     4.0.30319.42000                                                                                                                                             
WSManStackVersion              3.0                                                                                                                                                         
PSRemotingProtocolVersion      2.3                                                                                                                                                         
SerializationVersion           1.1.0.1                                                                                                                                                     

#### Version of the DSC module that was used

PS C:\Windows\system32> Get-InstalledModule

Version    Name                                Repository           Description                                                                                                            
-------    ----                                ----------           -----------                                                                                                            
4.2.0      SharePointDSC                       PSGallery            This DSC module is used to deploy and configure SharePoint Server 2013, 2016 and 2019, and covers a wide range of ar...
1.4.7      PackageManagement                   PSGallery            PackageManagement (a.k.a. OneGet) is a new way to discover and install software packages from around the web....

**Comment by ykuijs** [comment]
Hi @VerdonTrigance Do you have a little more info (like the Verbose logging) of the reason why the SPFarm resource fails? And are you 100% sure no database with the name "dev02_State" exists on the database server?

**Comment by VerdonTrigance** [comment]
> Hi @VerdonTrigance Do you have a little more info (like the Verbose logging) of the reason why the SPFarm resource fails? And are you 100% sure no database with the name "dev02_State" exists on the database server?

Yes, I'm definitely sure there is no such database in SQL Server. I reproduced it twice more before posting this. I placed verbose log above, so I'm not sure what do you mean by 'verbose logging'. Is there any other log exists? 
I gave up with DSC and installed Sharepoint manually. But may be in this month I will do installation for another VM and reproduce this again to collect more logs.

**Comment by ykuijs** [comment]
With Verbose logging I mean the logging that DSC outputs to the screen if you use the -Verbose parameter with the Start-DscConfiguration cmdlet. You can also find this logging by opening an **elevated** PowerShell window and running:
```PowerShell
Get-ChildItem -Path C:\Windows\System32\Configuration\ConfigurationStatus | Sort-Object -Property LastWriteTime | Select-Object -Last 10
```
Then check for the json file of the timestamp on which you ran the Start-DscConfiguration cmdlet. Use "notepad [filename]" from the same PowerShell window to open the file.

**Note:** Browsing to this folder using Explorer doesn't work most of the times, so you have to use the above command.

**Comment by ykuijs** [comment]
Fix implemented in next Bugfix PR

**Comment by ykuijs** [comment]
Fixed via PR #1226
